File: questions/902. Numbers At Most N Given Digit Set/num.cpp
class Solution {
public:
    int atMostNGivenDigitSet(vector<string>& digits, int n) {
        string digit = to_string(n);
        int nl = digit.size(), res = 0, dl = digits.size();
        for(int i = 1; i < nl; i++){
            res += pow(dl, i);
        }
        for(int i = 0; i < nl; i++){
            bool sameNumber = false;
            for(string& d : digits){
                if(d[0] < digit[i]){
                    res += pow(dl, nl - i - 1);
                }else if(d[0] == digit[i]){
                    sameNumber = true;
                }
            }
            if(!sameNumber) return res;
        }
        return res + 1;
    }
};

File: questions/902. Numbers At Most N Given Digit Set/num.java
class Solution {
    public int atMostNGivenDigitSet(String[] digits, int n) {
        String str = String.valueOf(n);
        int sl = str.length(), dl = digits.length, res = 0;
        for(int i = 1; i < sl; i++){
            res += Math.pow(dl, i);
        }
        for(int i = 0; i < sl; i++){
            boolean sameNumber = false;
            for(String s: digits){
                if(s.charAt(0) < str.charAt(i)){
                    res += Math.pow(dl, sl - i - 1);
                }else if(s.charAt(0) == str.charAt(i)){
                    sameNumber = true;
                }
            }
            if(!sameNumber) return res;
        }
        return res + 1;
    }
}
